### PR TITLE
AOT sync up with latest spec

### PR DESCRIPTION
**Pull request by xujuntwt95329**

### CODE REVIEW

**Comment by xujuntwt95329** [comment]
This is the AoT part of https://github.com/bytecodealliance/wasm-micro-runtime/pull/2836

File: tests/wamr-test-suites/spec-test-script/gc_ignore_cases.patch [hunk 1]
```diff
@@ -895,6 +906,36 @@ index 1ffd03f8..bdf7471f 100644
...
+diff --git a/test/core/return_call_ref.wast b/test/core/return_call_ref.wast
+index 353811f0..f49dc6e4 100644
+--- a/test/core/return_call_ref.wast
++++ b/test/core/return_call_ref.wast
+@@ -192,20 +192,20 @@
+ 
+ (assert_return (invoke "count" (i64.const 0)) (i64.const 0))
+ (assert_return (invoke "count" (i64.const 1000)) (i64.const 0))
+-(assert_return (invoke "count" (i64.const 1_000_000)) (i64.const 0))
++;; (assert_return (invoke "count" (i64.const 1_000_000)) (i64.const 0))
```
Comment (wenyongh) [inline tests/wamr-test-suites/spec-test-script/gc_ignore_cases.patch]: Could we change to 2000 or some smaller number

File: core/iwasm/aot/aot_loader.c [hunk 2]
```diff
@@ -1119,11 +1131,142 @@ destroy_table_init_data_list(AOTTableInitData **data_list, uint32 count)
...
+
+            if (field_count > 0) {
+                init_values->type_idx = type_idx;
+                init_values->count = field_count;
+
+                size = sizeof(WASMValue) * (uint64)field_count;
+                read_byte_array(buf, buf_end, init_values->fields, size);
+            }
+
+            expr->u.data = init_values;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c]: Should move this line to be after L1201? or in L1208, if read_byte_array fails, `wasm_runtime_free(expr->u.data)` may free an invalid memory.

File: core/iwasm/aot/aot_loader.c [hunk 3]
```diff
@@ -1119,11 +1131,142 @@ destroy_table_init_data_list(AOTTableInitData **data_list, uint32 count)
...
+
+                init_values->type_idx = type_idx;
+                init_values->length = length;
+
+                if (length > 0) {
+                    size = sizeof(WASMValue) * (uint64)length;
+                    read_byte_array(buf, buf_end, init_values->elem_data, size);
+                }
+
+                expr->u.data = init_values;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c]: Same as above

File: core/iwasm/aot/aot_loader.c [hunk 4]
```diff
@@ -1432,6 +1587,8 @@ load_types(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
             func_type->base_type.type_flag = type_flag;
             func_type->base_type.is_sub_final = is_sub_final;
             func_type->base_type.parent_type_idx = parent_type_idx;
+            func_type->base_type.rec_count = rec_count;
+            func_type->base_type.rec_idx = rec_idx;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c]: Seems that we can add a function to set these 5 fields, and call it for func_type/struct_type/array_type.

File: core/iwasm/aot/aot_loader.c [hunk 5]
```diff
@@ -1532,6 +1689,10 @@ load_types(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
             types[i] = (AOTType *)struct_type;
 
             struct_type->base_type.type_flag = type_flag;
+            struct_type->base_type.is_sub_final = is_sub_final;
+            struct_type->base_type.parent_type_idx = parent_type_idx;
+            struct_type->base_type.rec_count = rec_count;
+            struct_type->base_type.rec_idx = rec_idx;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c]: Same as above.

File: core/iwasm/aot/aot_loader.c [hunk 6]
```diff
@@ -1617,6 +1778,10 @@ load_types(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
             types[i] = (AOTType *)array_type;
 
             array_type->base_type.type_flag = type_flag;
+            array_type->base_type.is_sub_final = is_sub_final;
+            array_type->base_type.parent_type_idx = parent_type_idx;
+            array_type->base_type.rec_count = rec_count;
+            array_type->base_type.rec_idx = rec_idx;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c]: Same as above.

File: core/iwasm/aot/aot_loader.c [hunk 7]
```diff
@@ -1640,23 +1805,48 @@ load_types(const uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
...
         }
-        else {
-            module->types[i]->parent_type = NULL;
-            module->types[i]->root_type = module->types[i];
-            module->types[i]->inherit_depth = 0;
+
+        if ((rec_count == 0) || (rec_idx == rec_count - 1)) {
+            if (rec_count == 0) {
+                bh_assert(rec_idx == 0);
+            }
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c]: Duplicated code with L1808 to L1811, how about removing L1831 to L1835?

File: core/iwasm/compilation/aot_emit_aot_file.c [hunk 8]
```diff
@@ -263,10 +263,103 @@ get_mem_info_size(AOTCompData *comp_data)
...
+        case INIT_EXPR_TYPE_REFNULL_CONST:
+            /* type_index */
+            size += sizeof(uint32);
+            break;
+        case INIT_EXPR_TYPE_STRUCT_NEW_CANON:
+        {
+            WASMStructNewInitValues *struct_new_init_values =
+                (WASMStructNewInitValues *)expr->u.data;
+
+            // type_index + field_count + fields
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_aot_file.c]: How about `/* .. */`

File: core/iwasm/compilation/aot_emit_gc.c [hunk 9]
```diff
@@ -1456,6 +1457,156 @@ aot_compile_op_array_set(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
...
+    }
+    BUILD_ICMP(LLVMIntUGT, boundary, I32_CONST(UINT32_MAX), cmp[0],
+               "boundary_check1");
+    /* dst_offset + len > wasm_array_obj_length(dst_obj) */
+    if (!aot_array_obj_length(comp_ctx, array_obj, &array_len))
+        goto fail;
+    BUILD_ICMP(LLVMIntUGT, boundary, array_len, cmp[1], "boundary_check2");
+
+    /* logical or above 4 boundary checks */
+    for (i = 1; i < 2; ++i) {
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_gc.c]: Seems we can remove this line, and change below cmp[i] to cmp[1]

File: tests/wamr-test-suites/spec-test-script/gc_ignore_cases.patch (hunk 1 above)
Comment (xujuntwt95329) [reply tests/wamr-test-suites/spec-test-script/gc_ignore_cases.patch]: 1200 works, I'll change it

File: core/iwasm/aot/aot_loader.c (hunk 3 above)
Comment (xujuntwt95329) [reply core/iwasm/aot/aot_loader.c]: Done

File: core/iwasm/aot/aot_loader.c (hunk 4 above)
Comment (xujuntwt95329) [reply core/iwasm/aot/aot_loader.c]: Done

File: core/iwasm/aot/aot_loader.c (hunk 7 above)
Comment (xujuntwt95329) [reply core/iwasm/aot/aot_loader.c]: Done

File: core/iwasm/compilation/aot_emit_gc.c (hunk 9 above)
Comment (xujuntwt95329) [reply core/iwasm/compilation/aot_emit_gc.c]: Done

**Pull request merged**